src/Map.elm

```diff
@@ -138,92 +143,134 @@ type Action
-        ( landscapeAction, landscapeResources ) =
-            findSafeInList randomLandscapeResourceTypeId landscapeValue
-
-        ( resourceId, resourceSeed ) =
-            step (Random.int 0 (Array.length landscapeResources - 1)) resourceTypeSeed
+        ( _, landscapeResources ) =
+            findSafeInDict landscapeId l
+    in
+    genRandomLandscapeResource landscapeResources
+        |> Random.andThen
+            (\(LandscapeResource ( resourceAction, resources )) ->
+                genRandomResource resources
+                    |> Random.andThen
+                        (\(Resource ( ResourceName resourceName, Rarity rarity, maxAmount )) ->
+                            genAmountChance maxAmount
+                                |> Random.andThen
+                                    (\( amount, luck ) ->
+                                        if amount > 0 && succeed luck rarity then
+                                            let
+                                                r =
+                                                    ( Resource ( ResourceName resourceName, Rarity rarity, maxAmount )
+                                                    , resourceAction
+                                                    )
+                                            in
+                                            Random.constant (Just ( r, Amount amount ))
+
+                                        else
+                                            Random.constant
+                                                (Debug.log
+                                                    (stringifyLandscapeFailure resourceName amount luck)
+                                                    Nothing
+                                                )
+                                    )
+                        )
+            )
 
-        resource : Resource
-        resource =
-            findSafeInDict resourceId landscapeResources
 
-        ( resourceName, Rarity rarity, Amount maxAmount ) =
-            resource
+genRandomLandscapeResource : LandscapeResources -> Generator LandscapeResource
+genRandomLandscapeResource (LandscapeResources landscapeResources) =
+    Random.andThen
+        (\id ->
+            Random.constant (findSafeInList id landscapeResources)
+        )
+        (Random.int 0 (List.length landscapeResources - 1))
 
-        ( amount, amountSeed ) =
-            step (Random.int 0 maxAmount) resourceSeed
 
-        ( luck, luckSeed ) =
-            rollDice amountSeed
+genRandomResource : Resources Resource -> Generator Resource
+genRandomResource resources =
+    let
+        (Resources r) =
+            resources
     in
-    if amount > 0 && succeed luck rarity then
-        Just ( ( resource, landscapeAction ), Amount amount )
-
-    else
-        Debug.log ("Didn't find " ++ Debug.toString resource ++ "with amount#" ++ String.fromInt amount ++ " and luck#" ++ String.fromFloat luck) Nothing
+    Random.andThen
```

Comment:
Use `Random.map` instead, You only need to use `Random.andThen`  if you need to return another Generator.

Reply:
Ahh alright, will try. Thank you!